### DISCUSSION TITLE
configure custom backup handler (e.g. store backup in S3)

### DISCUSSION THREAD

**Discussion by lakedude** [Unleash/unleash#8160] [Q&A] [answered]
Hey everybody,

I'm starting to work with Gitlab's feature flags and the unleash java client. I found the option to bootstrap feature from an outside source like S3, but I don't see an option change the local backup. I would like the unleash client to store its backup file in the same S3 bucket, i.e. use the same file as for bootstrapping.

My concern is that on a multi instance scenario, all instances might initialise with the same features, but since each one might loose connection to Gitlab at any point, I would like to have a centralised backup file so every instance has the same feature state.

Thanks in advance!

**Comment by chriswk** [comment]
Hi @lakedude.

You can do this yourself by implementing the ToggleBootstrapProvider.

See https://github.com/unleash/unleash-client-java?tab=readme-ov-file#bootstrapping

**Reply by chriswk** [reply, marked as answer]
I see I need to upgrade my reading skills. Because you're asking for the write side of the backup to also be customizable. You'd need to look into a custom instance of FeatureRepository, which accepts a BackupHandler<File> (which exposes both write and read).

We don't want to have s3 as a dependency in the unleash sdk.

So work needed would be
1. Write an S3BackUpHandler which implements BackupHandler
2. Instantiate the FeatureRepository with your own backuphandler
`FeatureRepository fr = new FeatureRepository(unleashConfig, s3BackupHandler);`
3. Pass your FeatureRepository to DefaultUnleash's constructor
`new DefaultUnleash(UnleashConfig, fr, ...anyCustomStrategies)`

**Reply by lakedude** [reply]
thank you very much! This looks like exactly what I need. Using this, I probably won't need the `ToggleBootstrapS3Provider` as the backup handler checks S3 anyway, correct? So it would like something like this

```
val customBackupHandler = object : BackupHandler<FeatureCollection> {
    override fun write(collection: FeatureCollection) {
        TODO("write to S3")
    }

    override fun read(): FeatureCollection {
        TODO("read from S3")
    }
}
val unleashConfig = UnleashConfig.builder()
    .appName(appName)
    .instanceId(instanceId)
    .unleashAPI(api)
    .build()

unleashClient = DefaultUnleash(
    unleashConfig,
    FeatureRepository(unleashConfig, customBackupHandler)
)
```

**Reply by chriswk** [reply]
Yes, that should work. If it doesn't then please make a bug report :)
